README
NAME
    Data::Couplet - Yet another (But Hopefully Better) Key-Value Storage
    mechanism

VERSION
    version 0.02004313

SYNOPSIS
      use Data::Couplet;

      # Retain order.
      my $couplet = Data::Couplet->new(   a => $b , c => $d );

      my $output = $couplet->value('a');  # returns $b;

      my $hash = { 'this is a' => 'key' };

      $couplet->set( $hash, "hello");
      $couplet->value( $hash ); # hello

ALPHA CODE
    Lots of stuff is probably still broken, unimplemented, untested.

    User beware

DIFFERENT
    Why is this module different?

    1. No Tied Hashes.
        Tied hashes are IMO Ugly. Objects are far more handy for many
        things. Especially in moose world. You want tied hashes, do it
        yourself.

    2. Trying Hard to preserve non-scalar keys.
        I want it to be possible, to retain arbitrary references used as
        keys.

    3. Permutation.
        Its not here yet, but there *Will* eventually be reordering
        functions.

    I seriously looked all over CPAN for something that suited my needs and
    didn't find any.

    I then tried with Tie::IxHash::ButMoreFun, and then discovered that how
    I was using Tie::IxHash wasn't even sustainable on different versions of
    Perl, and based on the 1997 release date, I gave up on seeing that
    fixed.

METHODS
  CONSTRUCTOR
   ->new( %orderd_pairs )
    Create a new Data::Couplet entity using a series of ordered pairs.

      $c = Data::Couplet->new( 'a' => 'b', 'c' => 'd' );

  ENTRY CREATION
   ->set( Any $object, Any  $value ) : $self : Modifier
    Record the association of a key ( any object that can be coerced into a
    string ) to a value.

    New entries are pushed on the logical right hand end of it in array
    context.

      # { 'a' => 'b', 'c' => 'd' }
      set( 'a', 'e' );
      # { 'a' => 'e', 'c' => 'd' }
      set('e', 'a' );
      # { 'a' => 'e', 'c' => 'd', 'e' => 'a' }

  ENTRY REMOVAL
   ->unset( Array[Any] @objects ) : $self : Modifier
    Entries are ripped out of the structure, and all items moved around to
    fill the void.

      # { 'a' => 'b', 'c' => 'd','e'=>'f' }
      ->unset( 'c' );
      # { 'a' => 'b', 'e'=>'f' }
      ->unset('a');
      # { 'e' => 'f' }

   ->unset_at( Array[Int] @indices ) : $self : Modifier
    Like ->unset, except you know where ( logically ) in the order off
    things the entry you wish to delete is.

      ->unset_at( 1 );
      ->unset_at( 0 );

    Should be identical to the above code.

   ->unset_key( Array[Str] @keys ) : $self : Modifier
    This is what ->unset ultimately calls, except ->unset does implicit
    object_to_key conversion first. At present, that's not anything huge,
    its just $object to convert it to a string. But this may change at some
    future time. So use that method instead.

  VALUE MANIPULATION
   ->value( Any $object ) : Any $value
    Returns a value associated with a key object. See "unset" for the
    semantics of what object keys are.

   ->value_at( Int $index ) : Any $value
    Like value, but you need to know where in the data set the item is.

   ->values() : Any @list
    returns an array of all stored values in order.

   ->values_ref() : ArrayRef[Any] $list
    Just some nice syntax for [$o->values]

   ->key_values() : Any @list
    Returns an ordered sequence of key,value pairs, just like that passed to
    the constructor.

      my @d = $o->key_values()
      while( @d ){
        my $key = shift @d;
        my $value = shift @d;
        print "$key => $value\n"
      }

   ->key_values_paired() : Any[ArrayRef] @list
    Returns like ->key_values does but key/value is grouped for your
    convenience

      for ( $o->key_values_paired() ){
        my ( $key, $value ) = @{ $_ };
      }

  KEY MANIPULATION
   ->keys() : @list
    returns all known keys in order

   ->key_at( Int $index ) : String
    Given an index, return the key that holds that place.

   ->key_object( String $key ) : Any $object
    Given a string key, returns the object stored there.

    This is probably very unhelpful to you unless you explicitly asked us
    for our internal key name.

   ->key_object_at( Int $index ) : Any $object
    As with key_object, except partially useful, because you can fetch by
    ID.

  METHODS FROM PLUGINS
    By default, this package imports a few methods from various plug-ins.

    *   Data::Couplet::Plugin::KeyCount

        ->count
        ->last_index
        ->indices

    *   Data::Couplet::Plugin::BasicReorder

        ->move_up
        ->move_down
        ->swap

AUTHOR
    Kent Fredric <kentnl at cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Kent Fredric.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.02004314 2011-10-28T14:54:15Z
 [Dependencies]
 - changed: runtime requires Moose 0.84 -> 0.90
 - dropped: runtime requires MooseX::Has::Sugar

 [Features]
 - Added (private) accessors for everything moose can do on the ::Private
   core. This is soley to discourage consumers from accesing the hash
   internals directly to provide better encapsulation. Maybe one day
   ::Private will be entirely replaced with XS and you want it to remain
   working.
 - Added a basic Type Library, Data::Couplet::Types

0.02004313 2011-10-28T14:24:40Z
 - Maintenance release, no non-packaging changes in this release.

 [Dependencies::New]
 - Data::Dumper

 [Dependencies::Removed]
 - Test::use::ok
 - ok

 [Internals]
 - packages now set $VERSION without 'our'
 - packages now have $AUTHORITY

 [Packaging]
 - Much needed regeneration of Build.PL
 - Changes file migrated to CPAN::Changes format.
 - Copyright License updated ( Date, Indentation, Address )
 - No longer ships a Makefile.PL ( EUMM Sucks )
 - perlcriticrc -> perlcritic.rc
 - Moved to Git versioning
 - moved to Author::KENTNL
 - Repository metadata more featureful and uses https://
 - license is now 'perl_5' not 'perl'
 - dependencies now properly qualified.
 - now ship x_authority.
 - Pod overhaulled and rewoven.
 - Ship .perltidyrc
 - Tidy all code.

 [Tests]
 - added version reporting test.
 - t/* moved to xt/*
 - hand-written xt/* replaced by generated ones.
 - xt:added release/cpan-changes
 - xt:added release/distmeta
 - xt:added release/eol
 - xt:added release/kwalitee
 - xt:added release/pod-coverage
 - xt:added release/pod-syntax

0.02004312 2009-09-13T12:20:00Z
 - Moved last_id to last_index ( was a bad name anyway )
 - Added ->indices to the KeyCount plug. 
 - Added KeyPlug tests.
 - This is a minor release to tweak -last_index prior to S::F release

0.02004302 2009-09-13T02:06:26Z
 - Added a Ease-Of-Use subclassing utility.
 - Moved D::C core to the subclassing utility.
 - Added subclassing example.

0.02004222 2009-09-12T22:23:22Z
 - Added a plugin infrastructure w/ Roles.
 - Moved ->count to role composition ( you wont notice this )
 - Increased input strictness with really basic MX::Types based input
   validation. ( if you're going to be using _key or _at methods, use them
   properly ! :) )
 - Added ->last_id via Counter plug.
 - Moved the non-working sorting routines to a plug. Maybe they'll see some
   love there.

0.02004206 2009-09-12T06:01:08Z
 - Support for multiple-value commands ( Thanks James Laver )
 - Support for ->count() mechanism     ( Thanks James Laver )

0.02003813 2009-09-08T13:10:54Z
 - Clone Support works and tested.
 - Clones are MX::Cloned, and deep.
 - minor++ # version scheme incompatible
 - dist.ini now shipping with dist
 - DZIL CompileTests
 - Versioning Scheme switched to AutoVersion::Relative
 - Added MooseX::Clone support
 - Added MooseX::Types Deps
 - Added MooseX::Has::Sugar Deps
 - Added more author tests
 - Assload of Typos Fixed

0.0101 2009-08-22T10:09:57Z
 - Fixed some build deps

0.0100 2009-08-21T16:24:40Z
 - First version.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Data::Dump" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Scalar::Util" => 0,
    "Test::Moose" => 0,
    "Test::More" => "0.88"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Yet another (But Hopefully Better) Key-Value Storage mechanism",
  "dist_author" => [
    "Kent Fredric <kentnl at cpan.org>"
  ],
  "dist_name" => "Data-Couplet",
  "dist_version" => "0.02004313",
  "license" => "perl",
  "module_name" => "Data::Couplet",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Moose" => "0.84",
    "Moose::Role" => 0,
    "MooseX::Clone" => "0.04",
    "MooseX::Has::Sugar" => "0.0404",
    "MooseX::Types" => "0.19",
    "MooseX::Types::Moose" => "0.19",
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
